packages/tx/src/transaction.ts

```diff
@@ -21,290 +21,255 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
+      )
     }
 
-    // If the signature was verified successfully the _senderPubKey field is defined
-    return this._senderPubKey!
+    try {
+      return ecrecover(
+        msgHash,
+        this.v!.toNumber(),
+        bnToRlp(this.r),
+        bnToRlp(this.s),
+        this._signedTxImplementsEIP155() ? this.getChainId() : undefined,
+      )
+    } catch (e) {
+      throw new Error('Invalid Signature')
+    }
   }
 
   /**
    * Determines if the signature is valid
    */
   verifySignature(): boolean {
-    const msgHash = this.hash(false)
-    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
-    if (this._common.gteHardfork('homestead') && new BN(this.s).cmp(N_DIV_2) === 1) {
-      return false
-    }
-
     try {
-      const v = bufferToInt(this.v)
-      const useChainIdWhileRecoveringPubKey =
-        v >= this.getChainId() * 2 + 35 && this._common.gteHardfork('spuriousDragon')
-      this._senderPubKey = ecrecover(
-        msgHash,
-        v,
-        this.r,
-        this.s,
-        useChainIdWhileRecoveringPubKey ? this.getChainId() : undefined,
-      )
+      return unpadBuffer(this.getSenderPublicKey()).length !== 0
     } catch (e) {
       return false
     }
-
-    return !!this._senderPubKey
   }
 
   /**
-   * sign a transaction with a given private key
+   * Sign a transaction with a given private key
    * @param privateKey - Must be 32 bytes in length
    */
   sign(privateKey: Buffer) {
-    // We clear any previous signature before signing it. Otherwise, _implementsEIP155's can give
-    // different results if this tx was already signed.
-    this.v = Buffer.from([])
-    this.s = Buffer.from([])
-    this.r = Buffer.from([])
+    if (privateKey.length !== 32) {
+      throw new Error('Private key must be 32 bytes in length.')
+    }
 
-    const msgHash = this.hash(false)
-    const sig = ecsign(msgHash, privateKey)
+    const msgHash = this.getMessageToSign()
 
-    if (this._implementsEIP155()) {
-      sig.v += this.getChainId() * 2 + 8
+    let { v, r, s } = ecsign(msgHash, privateKey)
+
+    if (this._unsignedTxImplementsEIP155()) {
+      v += this.getChainId() * 2 + 8
     }
 
-    Object.assign(this, sig)
+    return new Transaction(
+      this.common,
+      this.nonce,
+      this.gasPrice,
+      this.gasLimit,
+      this.to,
+      this.value,
+      this.data,
+      new BN(v),
+      new BN(r),
+      new BN(s),
+    )
   }
 
   /**
    * The amount of gas paid for the data in this tx
    */
   getDataFee(): BN {
-    const data = this.raw[5]
     const cost = new BN(0)
-    for (let i = 0; i < data.length; i++) {
-      data[i] === 0
-        ? cost.iaddn(this._common.param('gasPrices', 'txDataZero'))
-        : cost.iaddn(this._common.param('gasPrices', 'txDataNonZero'))
+    for (let i = 0; i < this.data.length; i++) {
+      this.data[i] === 0
+        ? cost.iaddn(this.common.param('gasPrices', 'txDataZero'))
```

Comment:
Moving the param outside of the loop would be great. I believe is one of the performance hotspots that ethworks identified.

Also, can this sum be implemented on native numbers and then converted to BNs? If we assume that the `MAX_DATA_LENGTH * txDataNonZero` is < 2**53 that would be fine.

Reply:
nice idea, done in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/34bd63942cf766d9cf4ea5d649c27e613ee73d58